botorch/optim/optimize_mixed.py

```diff
@@ -164,10 +164,75 @@ def get_nearest_neighbors(
     return unique_neighbors
 
 
+def get_categorical_neighbors(
+    current_x: Tensor,
+    bounds: Tensor,
+    cat_dims: Tensor,
+    max_num_cat_values: int = MAX_DISCRETE_VALUES,
+) -> Tensor:
+    r"""Generate all 1-Hamming distance neighbors of a given input. The neighbors
+    are generated for the categorical dimensions only.
+
+    We assume that all categorical values are equidistant. If the number of values
+    is greater than `max_num_cat_values`, we sample uniformly from the range of
+    possible values for that dimension.
+
+    NOTE: This assumes that `current_x` is detached and uses in-place operations,
+    which are known to be incompatible with autograd.
+
+    Args:
+        current_x: The design to find the neighbors of. A tensor of shape `d`.
+        bounds: A `2 x d` tensor of lower and upper bounds for each column of `X`.
+        cat_dims: A tensor of indices corresponding to categorical parameters.
+        max_num_cat_values: Maximum number of values for a categorical parameter,
+            beyond which values are uniformly sampled.
+
+    Returns:
+        A tensor of shape `num_neighbors x d`, denoting all unique 1-Hamming
+        distance neighbors.
+    """
+
+    # Neighbors are generated by considering all possible values for each
+    # categorical dimension, one at a time.
+    def _get_cat_values_tensor(lb: Tensor, ub: Tensor) -> Tensor:
+        r"""Get a tensor of all of the values a categorical feature may take."""
+        if ub - lb + 1 <= max_num_cat_values:
+            return torch.arange(
+                lb, ub + 1, dtype=current_x.dtype, device=current_x.device
+            )
+        else:
+            return (
```

Comment:
As codecov says, this is currently not being tested - can you please add a test for the case with too many categorical values as well?